File: frontend/src/types/events.ts
import type { UiComponent } from "./components";

export const SUPPORTED_PROTOCOL_VERSION = "1" as const;

export type AgentEventType =
  | "status"
  | "text_delta"
  | "text"
  | "tool_start"
  | "tool_result"
  | "final_result"
  | "error"
  | "done";

export interface AgentEvent {
  protocol_version: typeof SUPPORTED_PROTOCOL_VERSION;
  type: AgentEventType | (string & {});
  content?: string | null;
  tool_name?: string | null;
  tool_call_id?: string | null;
  ui_component?: UiComponent | null;
  metadata?: Record<string, unknown>;
}

export function isUiComponent(value: unknown): value is UiComponent {
  if (!value || typeof value !== "object" || Array.isArray(value)) return false;
  const type = (value as { type?: unknown }).type;
  return (
    type === "dataframe" ||
    type === "table_list" ||
    type === "table_schema" ||
    type === "final_result"
  );
}

File: src/sql_agent/core/events.py
"""Simple text / event stream models for Agent output."""

from typing import Any, Dict, Literal, Optional

from pydantic import BaseModel, Field

from sql_agent.components import UiComponent

PROTOCOL_VERSION: Literal["1"] = "1"


class AgentEvent(BaseModel):
    """Lightweight event yielded by Agent.send_message.

    type 常用值:
      - status: 状态提示
      - text_delta: 流式文本增量
      - text: 完整文本回复
      - tool_start: 开始执行工具
      - tool_result: 工具执行结果
      - final_result: 经校验的最终分析结果
      - error: 错误
      - done: 本轮结束
    """

    protocol_version: Literal["1"] = PROTOCOL_VERSION
    type: str = Field(description="Event type")
    content: Optional[str] = Field(default=None, description="Text payload")
    tool_name: Optional[str] = Field(default=None)
    tool_call_id: Optional[str] = Field(default=None)
    ui_component: Optional[UiComponent] = Field(default=None)
    metadata: Dict[str, Any] = Field(default_factory=dict)
